src/Strategy/Pub.hs

```diff
@@ -57,8 +61,8 @@ mkProject project =
 getDeps :: (Has Exec sig m, Has ReadFS sig m, Has Diagnostics sig m, Has Logger sig m) => PubProject -> m DependencyResults
 getDeps project = do
   (graph, graphBreadth) <- case pubLock project of
-    Just lockFile -> analyzeDepsCmd lockFile (pubSpecDir project) <||> analyzePubLockFile lockFile
-    Nothing -> analyzePubSpecFile $ pubSpec project
+    Just lockFile -> analyzeDepsCmd lockFile (pubSpecDir project) <||> (warnOnSuccess PubspecLimitation $ analyzePubLockFile lockFile)
+    Nothing -> warnOnSuccess PubspecLimitation $ analyzePubSpecFile $ pubSpec project
```

Comment:
@cnr can you have a look at this pattern?

I don't think warnOnErr pattern as described in devdocs is the best way forward, 
```haskell
mavenAnalyzer project = context "Maven" $
  mvnCommandAnalysis project <||> mvnStaticAnalysis project

mvnCommandAnalysis project =
  context "Dynamic analysis"
    . warnOnErr MissingDeps
    . warnOnErr MissingEdges
    $ MvnDepTree.analyze project

mvnStaticAnalysis project =
  context "Static analysis"
    $ MvnPom.analyze project
```

I prefer:
```haskell
mavenAnalyzer project = context "Maven" $
  mvnCommandAnalysis project <||> mvnStaticAnalysis project

mvnCommandAnalysis project =
  context "Dynamic analysis"
    $ MvnDepTree.analyze project

mvnStaticAnalysis project =
  context "Static analysis"
    . warnOnSuccess MissingDeps
    . warnOnSuccess MissingEdges
    $ MvnPom.analyze project
```

Reply:
I know it seems backwards. Here's why it's `warnOnErr` rather than `warnOnSuccess`:

A `warnOnSuccess` mechanism would allow us to emit standalone warnings like "Missing deps" or "Missing edges", but it's missing something I think is critical: the underlying context for why the warning was emitted (i.e., the underlying error or errors).

In the first example with `warnOnErr`, missing deps and missing edges are _caused by the failure of_ `MvnDepTree.analyze`. We can then embed the error details inside of the warning (and these details are displayed alongside the warning when rendered)

Contrast with the second example, where we know we're missing deps and missing edges, but we don't know _why_

---

while in practice I think `warnOnSuccess` could be clearer in some cases, having that underlying context for the warning is really important, and that's why I didn't include it

Reply:
as a sidenote, `warnOnSuccess` can be implemented in terms of the existing operations, though again _I do not recommend this_:

```hs
warnOnSuccess :: (ToDiagnostic warn, Has Diagnostics sig m) => warn -> m a -> m a
warnOnSuccess w m = do
  res <- m
  warn w
  pure res
```

this takes advantage of the short-circuiting behavior of `>>=` when a `Failure` is encountered

Reply:
one final note: something I documented poorly is the intended semantic meanings of warnings vs error context:

- warnings are in general meant to be a simple consequence of some failure (i.e. a general "the analysis may be missing dependencies")
- error context is meant to decorate _why_ an error happened (incl documentation and how to troubleshoot)

this is reflected in how they're currently rendered -- the example from that PR I think illustrates this best is this one: ![](https://user-images.githubusercontent.com/3586257/151616747-bd3bc5db-f75c-4168-aac7-4c576fd09824.png)

where `Analysis result may be missing dependencies` comes from the `warnOnErr` warning, and the message in the "Details" section was added via `errCtx` on the underlying error:

```hs
foo dir = do
  res <-
    recover
      . warnOnErr MissingDeps
      . errCtx (PipenvCmdFailureCtx dir)
      $ PipenvCmd.doTheThing dir
```

Reply:
Thanks connor - this is great feedback 👍